ipa-core/src/protocol/hybrid/agg.rs

```diff
@@ -0,0 +1,524 @@
+            build_prf_hybrid_report(31, 1, 2),  // triple
+            build_prf_hybrid_report(31, 3, 4),  // triple
+            build_prf_hybrid_report(31, 5, 6),  // triple
+        ];
+
+        let expected = vec![
+            [
+                build_aggregateable_report(4, 0),
+                build_aggregateable_report(0, 13),
+            ],
+            [
+                build_aggregateable_report(2, 0),
+                build_aggregateable_report(0, 3),
+            ],
+        ];
+
+        let results = group_report_pairs_ordered(reports);
+        assert_eq!(results, expected);
+    }
+
+    /// This test checks that the sharded malicious `aggregate_reports` fails
+    /// under a simple bit flip attack by H1.
+    #[test]
+    #[should_panic(expected = "DZKPValidationFailed")]
+    fn sharded_fail_under_bit_flip_attack_on_breakdown_key() {
+        run_random(|mut rng| async move {
+            let target_shard = ShardIndex::from(rng.gen_range(0..u32::try_from(SHARDS).unwrap()));
+            let world = TestWorld::<WithShards<SHARDS>>::with_shards(TestWorldConfig::default());
+            let records = get_records();
+            let _results: Vec<[Vec<AggregateableHybridReport<BA8, BA3>>; 3]> = world
+                .malicious(records.clone().into_iter(), |ctx, input| {
+                    let match_keys = match ctx.shard_id() {
+                        ShardIndex(0) => SHARD1_MKS,
+                        ShardIndex(1) => SHARD2_MKS,
+                        _ => panic!("invalid shard_id"),
+                    };
+                    async move {
+                        let indistinguishable_reports: Vec<
+                            IndistinguishableHybridReport<BA8, BA3>,
+                        > = input.iter().map(|r| r.clone().into()).collect::<Vec<_>>();
+
+                        let mut prf_reports: Vec<PrfHybridReport<BA8, BA3>> =
+                            indistinguishable_reports
+                                .iter()
+                                .zip(match_keys)
+                                .map(|(indist_report, match_key)| PrfHybridReport {
+                                    match_key,
+                                    value: indist_report.value.clone(),
+                                    breakdown_key: indist_report.breakdown_key.clone(),
+                                })
+                                .collect::<Vec<_>>();
+
+                        // flip a bit of the match_key on the target shard, H1
+                        if ctx.shard_id() == target_shard && ctx.role() == Role::H1 {
+                            prf_reports[0].breakdown_key +=
+                                Replicated::new(BA8::ZERO, BA8::truncate_from(1_u128));
+                        }
+
+                        aggregate_reports(ctx.clone(), prf_reports).await.unwrap()
+                    }
+                })
+                .await;
+        });
+    }
```

Comment:
I wasn't able to get the intercepter to work, but I'm able to manually XOR one of the shares, and that causes the expected failure. cc @akoshelev @andyleiserson